latch_config/config.py

```diff
@@ -51,17 +70,92 @@ class LoggingMode(str, Enum):
 T = TypeVar("T")
 
 
-def read_config(x: type[T], env_prefix: str = "") -> T:
+class _IsDataclass(Protocol):
+    __dataclass_fields__: ClassVar[dict[str, Field]]
+
+
+DC = TypeVar("DC", bound=_IsDataclass)
+
+
+class NoValue: ...
+
+
+class NestyError(Exception):
+    def __init__(self, msg: str, children: list[Self] | None = None):
+        self.msg = msg
+        self.children = children
+
+    def __str__(self):
+        if self.children is None:
+            return self.msg
+
+        res: list[str] = [self.msg]
+        for child in self.children:
+            res.append(indent(str(child), " -> "))
+
+        return "\n".join(res)
+
+
+def parse_config_val(t: type[T], v: Any) -> T:
+    if t is type(None):
+        if v is None:
+            return v
+
+        raise NestyError(f"value `{repr(v)}` is not assignable to `{t}`")
+
+    origin, args = get_origin(t), get_args(t)
+
+    if origin is None:
+        try:
+            if not isinstance(v, t):
```

Comment:
move this outside of the try-catch

Reply:
also definitely constrain this to just primitives - calling random class constructors could have side-effects